arrow-udf-python/src/pyarrow.rs

```diff
@@ -134,103 +65,223 @@ macro_rules! build_array {
     }};
 }
 
-/// Build arrow array from python objects.
-pub fn build_array(field: &Field, py: Python<'_>, values: &[PyObject]) -> PyResult<ArrayRef> {
-    match field.data_type() {
-        DataType::Null => build_array!(NullBuilder, py, values),
-        DataType::Boolean => build_array!(BooleanBuilder, py, values),
-        DataType::Int8 => build_array!(Int8Builder, py, values),
-        DataType::Int16 => build_array!(Int16Builder, py, values),
-        DataType::Int32 => build_array!(Int32Builder, py, values),
-        DataType::Int64 => build_array!(Int64Builder, py, values),
-        DataType::UInt8 => build_array!(UInt8Builder, py, values),
-        DataType::UInt16 => build_array!(UInt16Builder, py, values),
-        DataType::UInt32 => build_array!(UInt32Builder, py, values),
-        DataType::UInt64 => build_array!(UInt64Builder, py, values),
-        DataType::Float32 => build_array!(Float32Builder, py, values),
-        DataType::Float64 => build_array!(Float64Builder, py, values),
-        DataType::Utf8 => match field
-            .metadata()
-            .get("ARROW:extension:name")
-            .map(|s| s.as_str())
-        {
-            Some("arrowudf.json") => {
-                let json_dumps = py.eval_bound("json.dumps", None, None)?;
-                let mut builder = StringBuilder::with_capacity(values.len(), 1024);
-                for val in values {
-                    if val.is_none(py) {
-                        builder.append_null();
-                        continue;
-                    };
-                    let json_str = json_dumps.call1((val,))?;
-                    builder.append_value(json_str.extract::<&str>()?);
+macro_rules! build_json_array {
+    ($py:expr, $pyobjects:expr) => {{
+        let json_dumps = $py.eval_bound("json.dumps", None, None)?;
+        let mut builder = StringBuilder::with_capacity($pyobjects.len(), 1024);
+        for pyobj in $pyobjects {
+            if pyobj.is_none($py) {
+                builder.append_null();
+                continue;
+            };
+            let json_str = json_dumps.call1((pyobj,))?;
+            builder.append_value(json_str.extract::<&str>()?);
+        }
+        Ok(Arc::new(builder.finish()))
+    }};
+}
+
+#[derive(Debug, Clone)]
+pub struct Converter {
+    arrow_extension_key: String,
+    json_extension_name: String,
+    decimal_extension_name: String,
+}
+
+impl Converter {
+    pub fn new() -> Self {
+        Self {
+            arrow_extension_key: "ARROW:extension:name".to_string(),
+            json_extension_name: "arrowudf.json".to_string(),
+            decimal_extension_name: "arrowudf.decimal".to_string(),
+        }
+    }
+
+    // TODO: learn how to implement with_ pattern
+    // also add this to all modules
+    #[allow(dead_code)]
+    pub fn set_arrow_extension_key(&mut self, key: &str) {
+        self.arrow_extension_key = key.to_string();
+    }
```

Comment:
something like this?
```suggestion
    #[allow(dead_code)]
    pub fn with_arrow_extension_key(mut self, key: &str) -> Self {
        self.arrow_extension_key = key.to_string();
        self
    }
```

Reply:
actually I'm thinking that the set_ functions are more practical after having integrated this into Databend.

Reply:
either way is fine to me.